### ISSUE TITLE
`nc` doesnt' find system certificate store anymore

### ISSUE THREAD

**Issue by andir** [NixOS/nixpkgs#68286]
**Describe the bug**
The `nc` attribute (provided by LibreSSL) is no longer able to find the system certificates on NixOS. It seems like with the change to the CMake build system / with the switch to libressl 2.9 the feature broke.

**To Reproduce**
1. `nix-shell --pure -p nc` with a current master / recent unstable channel.
2. `nc -cvz google.com 443`
3. See `nc: failed to open CA file '/etc/ssl/cert.pem': No such file or directory`

**Expected behavior**
`nc` should have been able to connect to find the local system certificates and connect successfully.

**Metadata**

 - system: `"x86_64-linux"`
 - host os: `Linux 5.2.11, NixOS, 19.09.git.920dd87 (Loris)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.3pre6895_84de821`
 - channels(root): `"nixos-19.09pre190978.8d1510abfb5"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`

Maintainer information:
```yaml
# a list of nixpkgs attributes affected by the problem
attribute:
  - nc
  - libressl_2_9.nc
  - libressl_3_0.nc
# a list of nixos modules affected by the problem
module: []
```
cc libressl maintainers @fpletz @thoughtpolice

**Comment by d-goldin** [comment]
I ran a bisect and it seems like dc193761a8ce3cb44aacffad76d47159d505429b broke it. I tested it with libressl_2_8 and libressl_3_0 additionally and those seems to be fine.

I toyed around a little bit and a patch like below seems to fix the issue for this particular version (they changes parts of this mechanism between 2.8 and 2.9).

```
diff --git a/pkgs/development/libraries/libressl/default.nix b/pkgs/development/libraries/libressl/default.nix
index 71b8d2bb20e..7fe19ba097a 100644
--- a/pkgs/development/libraries/libressl/default.nix
+++ b/pkgs/development/libraries/libressl/default.nix
@@ -21,6 +21,7 @@ let
       # .note.GNU-stack section, and if that section is missing from any object,
       # the linker will make the stack executable.
       "-DCMAKE_C_FLAGS=-DHAVE_GNU_STACK"
+      "-DDEFAULT_CA_FILE=$out/etc/ssl/cert.pem"
     ];
 
     # The autoconf build is broken as of 2.9.1, resulting in the following error
```

**Comment by andir** [comment]
That patch does nothing for me. `2.8` just works as it currently is in nixpkgs. `2.9` & `3.0` show no change in my mentioned test. How did you verify?

**Comment by d-goldin** [comment]
@andir: you're right, that doesn't fix it. Must have accidentally invoked some non broken version. I'll give it another look, out of curiosity.

**Comment by d-goldin** [comment]
So, having given it a closer look again - `2.9` and `3.0` are both equally broken.
There was a change made, where formerly there was the following in `apps/nc/netcat.c`:

```
[...]
12 #ifndef DEFAULT_CA_FILE
13 #define DEFAULT_CA_FILE "/etc/ssl/cert.pem"
14 #endif
[...]
```

Now, in `2.9` and `3.0` it was changed to simply pre-initialize using `tls_default_ca_cert_file()` from `tls/tls_config.c` which just returns a constant location of `/etc/ssl/cert.pem` so seems like there is no way override it through build config or macros anymore. That would likely mean we'd have to patch the default location from `tls_config.c`.

**Comment by d-goldin** [comment]
So, a rather quick and dirty patch that seems to work is:

```
diff --git a/pkgs/development/libraries/libressl/default.nix b/pkgs/development/libraries/libressl/default.nix
index 71b8d2bb20e..89e24edf5dc 100644
--- a/pkgs/development/libraries/libressl/default.nix
+++ b/pkgs/development/libraries/libressl/default.nix
@@ -28,6 +28,9 @@ let
     # Fortunately LibreSSL provides a CMake build as well, so opt for CMake by
     # removing ./configure pre-config.
     preConfigure = ''
+      # Since 2.9.x the default location can't be configured from the build using
+      # DEFAULT_CA_FILE anymore, instead we have to patch the default value.
+      sed -i "s|\"/etc/ssl/cert.pem\"|\"$out/etc/ssl/cert.pem\"|" ./tls/tls_config.c;
       rm configure
     '';
 ```

**Comment by andir** [comment]
@d-goldin That doesn't sound too bad. Do you mind opening a PR for that? You can use `lib.versionAtLeast` to gate it behind a version switch?

**Comment by d-goldin** [comment]
@andir: Will do a bit later. Have you had a chance to test this one? Just to make sure.

**Comment by d-goldin** [comment]
There is another PR for this too: https://github.com/NixOS/nixpkgs/pull/68456

**Comment by d-goldin** [comment]
@andir: The issue should be fixed now on master and release-19.09.

**Comment by andir** [comment]
Confirmed. Fixed on master :+1: 

Thank you for your working on this!
